control-plane/connect-inject/controllers/pod/pod_controller.go

```diff
@@ -210,33 +233,238 @@ func (r *Controller) writeWorkload(ctx context.Context, pod corev1.Pod) error {
 	}
 
 	// TODO(dans): replace with common.ToProtoAny when available
-	proto, err := anypb.New(workload)
+	data, err := anypb.New(workload)
 	if err != nil {
 		return fmt.Errorf("could not serialize workload: %w", err)
 	}
 
-	// TODO: allow custom workload metadata
-	meta := map[string]string{
-		constants.MetaKeyKubeNS: pod.Namespace,
-		metaKeyManagedBy:        constants.ManagedByPodValue,
+	req := &pbresource.WriteRequest{
+		Resource: &pbresource.Resource{
+			Id:       getWorkloadID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
+			Metadata: metaFromPod(pod),
+			Data:     data,
+		},
+	}
+	_, err = r.ResourceClient.Write(ctx, req)
+	return err
+}
+
+func (r *Controller) writeProxyConfiguration(ctx context.Context, pod corev1.Pod) error {
+
+	mode, err := r.getTproxyMode(ctx, pod)
+	if err != nil {
+		return fmt.Errorf("failed to get transparent proxy mode: %w", err)
+	}
+
+	exposeConfig, err := r.getExposeConfig(pod)
+	if err != nil {
+		return fmt.Errorf("failed to get expose config: %w", err)
+	}
+
+	bootstrapConfig, err := r.getBootstrapConfig(pod)
+	if err != nil {
+		return fmt.Errorf("failed to get bootstrap config: %w", err)
+	}
+
+	if exposeConfig == nil &&
+		bootstrapConfig == nil &&
+		mode == pbmesh.ProxyMode_PROXY_MODE_DEFAULT {
+		// It's possible to remove interesting annotations and need to clear any existing config,
+		// but for now we treat pods as immutable configs owned by other managers.
+		return nil
+	}
+
+	pc := &pbmesh.ProxyConfiguration{
+		Workloads: &pbcatalog.WorkloadSelector{
+			Names: []string{pod.GetName()},
+		},
+		DynamicConfig: &pbmesh.DynamicConfig{
+			Mode:         mode,
+			ExposeConfig: exposeConfig,
+		},
+		BootstrapConfig: bootstrapConfig,
+	}
+
+	// TODO(dans): replace with common.ToProtoAny when available
+	data, err := anypb.New(pc)
+	if err != nil {
+		return fmt.Errorf("could not serialize workload: %w", err)
 	}
 
 	req := &pbresource.WriteRequest{
 		Resource: &pbresource.Resource{
-			Id:       getWorkloadID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
-			Metadata: meta,
-			Data:     proto,
+			Id:       getProxyConfigurationID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
+			Metadata: metaFromPod(pod),
+			Data:     data,
 		},
 	}
 	_, err = r.ResourceClient.Write(ctx, req)
 	return err
 }
 
-//func (r *Controller) writeHealthStatus(pod corev1.Pod) error {
-//	return nil
-//}
+func (r *Controller) getTproxyMode(ctx context.Context, pod corev1.Pod) (pbmesh.ProxyMode, error) {
+	// A user can enable/disable tproxy for an entire namespace.
+	var ns corev1.Namespace
+	err := r.Client.Get(ctx, types.NamespacedName{Name: pod.GetNamespace()}, &ns)
+	if err != nil {
+		return pbmesh.ProxyMode_PROXY_MODE_DEFAULT, fmt.Errorf("could not get namespace info for %s: %w", pod.GetNamespace(), err)
+	}
```

Comment:
~ Should we plumb the ns through higher up to simplify testing and so we don't accidentally introduce redundant fetches later in other parts of the code?

Reply:
I think it's a good idea, but I'll defer optimizing when we need this data in some other context.